### ISSUE TITLE
[Bug]: Type mismatch in EcommerceBundle installed classes

### ISSUE THREAD

**Issue by jremmurd** [pimcore/pimcore#11063]
### Expected behavior

Installed classes of EcommerceBundle match the provided abstract classes. 

See i.e. https://github.com/pimcore/pimcore/blob/10.x/bundles/EcommerceFrameworkBundle/Resources/install/class_sources/class_OnlineShopOrderItem_export.json#L289

vs. 

https://github.com/pimcore/pimcore/blob/10.x/bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php#L72

### Actual behavior

An error occures

` Compile Error: Declaration of Pimcore\Model\DataObject\OnlineShopOrderItem::getTotalPrice(): ?float must be compatible with Pimcore\Bundle\EcommerceFrameworkBundle\Model\AbstractOrderItem::getTotalPrice(): ?string`

### Steps to reproduce

Create OnlineShopOrderItem and extend from AbstractOrderItem of EcommerceBundle.

See https://github.com/pimcore/pimcore/blob/10.x/bundles/EcommerceFrameworkBundle/Resources/install/class_sources/class_OnlineShopOrderItem_export.json#L289

**Comment by jremmurd** [comment]
The Issue only occurs if the settings for the number field in the OnlineShopOrder Class are changed. If there are max. decimals defined the type changes from float to string and the definitions are compatible.

**Comment by ThisIsJustARandomGuy** [comment]
Can we reopen this? I didn't change any of the values from the default and they were still empty. So I got the same error after trying to upgrade (in addition to #10910). The site was originally running on 6.8 but has been on 6.9 for a while now.

**Comment by teklakct** [comment]
This is "silly". 
In [Upgrade_Notes for v10.0.0](https://pimcore.com/docs/pimcore/current/Development_Documentation/Installation_and_Upgrade/Upgrade_Notes/index.html) we can read that:
> [Ecommerce] Changed price fields `totalNetPrice` and `totalPrice` of `OnlineShopOrderItem` to decimal.

In the generated class definition for DataObject `OnlineShopOrderItem` typings are correct but in the extended class but in the [AbstractOrderItem](https://github.com/pimcore/pimcore/blob/v10.5.17/bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php#L82) from Ecommerce framework, you still have wrong typing. 

Really annoying that your code is not compatible on its own.

**Comment by jdreesen** [comment]
Why don't you provide a PR to fix it?

**Comment by teklakct** [comment]
I'm new to Pimcore and really don't know how it exactly works. I can prepare a MR with the changes of typing in  `[AbstractOrderItem](https://github.com/pimcore/pimcore/blob/v10.5.17/bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php#L82)` but I'm not sure if this is enough. 
@jdreesen if you can be my wingman I can do even in upcoming minutes.

**Comment by jdreesen** [comment]
Actually, I'm pretty sure the return type of `AbstractOrderItem::getTotalNetPrice(): ?string` is correct. Are you sure your `totalNetPrice` field is a `number` type with `decimalSize` and `decimalPrecision` configured? If they're not set, the type will be `float`, if they are set, it will be `string`.

![grafik](https://user-images.githubusercontent.com/424602/221702880-23da18b5-b72a-4a52-a9cd-d7b0ceee2ae4.png)

**Comment by teklakct** [comment]
Thanks. In my setup decimal size and precision are defined
<img width="686" alt="Screenshot 2023-02-28 at 00 10 33" src="https://user-images.githubusercontent.com/1399832/221710247-39aa9b4a-fb21-42fd-a0bd-8adf1f718531.png">

and the generated class methods are typed as following:

```
public function getTotalNetPrice(): ?float
public function setTotalNetPrice(?float $totalNetPrice)
public function getTotalPrice(): ?float
public function setTotalPrice(?float $totalPrice)
```

So the generated class works perfectly find, but the my problem is related with the Abstract class
```
class OnlineShopOrderItem extends \Pimcore\Bundle\EcommerceFrameworkBundle\Model\AbstractOrderItem
```
where in [AbstractOrderItem](https://github.com/pimcore/pimcore/blob/v10.5.17/bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php#L82) we have

```
    /**
     * @return string|null
     */
    abstract public function getTotalPrice(): ?string;

    /**
     * @param string|null $totalPrice
     */
    abstract public function setTotalPrice(?string $totalPrice);

    /**
     * @return string|null
     */
    abstract public function getTotalNetPrice(): ?string;

    /**
     * @param string|null $totalNetPrice
     */
    abstract public function setTotalNetPrice(?string $totalNetPrice);
```

I was surprised that even in [OrderManager](https://github.com/pimcore/pimcore/blob/v10.5.17/bundles/EcommerceFrameworkBundle/OrderManager/V7/OrderManager.php#L492) there is a `TODO` about the price type.

I try to find a way how to check if there a string or float is used in OrderManager

**Comment by teklakct** [comment]
@jdreesen can you guide me if there could be sth in my field configuration or in code? I am blink right now. 

And, I did not write it before. I'm using Pimcore v10.5.17 and use MariaDB 10.5 if this change anything

**Comment by teklakct** [comment]
I found that in my `definition_OnlineShopOrderItem.php` the Fieldset children list of `Price Information` the `totalNetPrice` and `totalPrice` are defined as `Pimcore\Model\DataObject\ClassDefinition\Data\Numeric` with `'fieldtype' => 'numeric',` 
I am not sure why this is happening. Will try to trace it and will let you know.

**Comment by jdreesen** [comment]
"Numeric" should be right. The generated scalar types depend on the config, this is where they come from: https://github.com/pimcore/pimcore/blob/10.5/models/DataObject/ClassDefinition/Data/Numeric.php#L152-L163

**Comment by teklakct** [comment]
@jdreesen 🎖️ I found it. Thank you a lot !!!! 

tl;dr
Data migration to Pimcore X failed. 

---
So what happened? 

After migration settings for decimal size and precision of `totalNetPrice` and `totalPrice` in `OnlineShopOrderItem` was missing (`null`). Because of that phpType was wrong.
Also, decimal size and precision **were** defined for `amount`, but should be `null`. 
Because of that class was generated wrongly. 

## How to fix that 
Maybe it could be useful for someone.

> _must be confirmed by works for me_
> 

### Option one -> in code

1. Open `definition_OnlineShopOrderItem.php` and find definitions for `totalNetPrice` and `totalPrice`
2. Setup decimal size and precision, or defined it if missing:
e.g.
```
Pimcore\Model\DataObject\ClassDefinition\Data\Numeric::__set_state(array(
     'name' => 'totalPrice',
     'title' => 'Price',
     // ...
     'decimalSize' => 19,
     'decimalPrecision' => 4,
```
3. Rebuild db structure for classes `bin/console pimcore:deployment:classes-rebuild`
4. Rebuilds php files for classes  `bin/console pimcore:build:classes`

### Option two -> in admin

1. Open `Setting` -> `Data Objects` -> `Classes` -> `E-commerce` and find `totalNetPrice` and `totalPrice` in `PriceInformation`. 
2. In "Specific settings" setup decimal size and precision 
e.g.
![Screenshot 2023-02-28 at 15 37 56](https://user-images.githubusercontent.com/1399832/221885881-38dba8c4-31cc-4b8c-ada9-5deaefda3c19.png)
3. Save changes
5. Rebuilds php files for classes  `bin/console pimcore:build:classes`
